core/schema/cache.go

```diff
@@ -32,13 +34,68 @@ func (s *cacheSchema) Dependencies() []SchemaResolvers {
 }
 
 type cacheArgs struct {
-	Key string
+	Key       string
+	Namespace string `default:""`
 }
 
-func (s *cacheSchema) cacheVolume(ctx context.Context, parent *core.Query, args cacheArgs) (*core.CacheVolume, error) {
-	// TODO(vito): inject some sort of scope/session/project/user derived value
-	// here instead of a static value
-	//
-	// we have to inject something so we can tell it's a valid ID
-	return core.NewCache(args.Key), nil
+func (s *cacheSchema) cacheVolume(ctx context.Context, parent dagql.Instance[*core.Query], args cacheArgs) (dagql.Instance[*core.CacheVolume], error) {
+	var inst dagql.Instance[*core.CacheVolume]
+
+	if args.Namespace != "" {
+		return dagql.NewInstanceForCurrentID(ctx, s.srv, parent, core.NewCache(args.Key+args.Namespace))
+	}
+
+	m, err := parent.Self.Server.CurrentModule(ctx)
+	if err != nil && !errors.Is(err, core.ErrNoCurrentModule) {
+		return inst, err
+	}
+
+	namespaceKey := ""
+	if m != nil {
+		name, err := m.Source.Self.ModuleName(ctx)
+		if err != nil {
+			return inst, err
+		}
+
+		bk, err := m.Source.Self.Query.Buildkit(ctx)
+		if err != nil {
+			return inst, err
+		}
+
+		refString, err := m.Source.Self.RefString()
+		if err != nil {
+			return inst, err
+		}
+		parsedRef := parseRefString(ctx, bk, refString)
+
+		// use combination of name, modPath and subpath as namespace key.
+		// if we only use name there is a high chance for namespace key conflict
+		// and high chance of cache-miss if we use module source digest for namespace key
+		namespaceKey = name + parsedRef.modPath + parsedRef.repoRootSubdir
```

Comment:
We can just use `m.Source.Self.Symbolic` right? That includes the clone-ref + subdir without the version information.

`parseRefString` is actually potentially an expensive call - it potentially makes network requests to try and determine the root/etc. We've already got that information available in the parsed module source, so we should prefer that.

Reply:
yep, this seems to have worked as you suggested 👍